File: apps/desktop/tests/limits.test.ts
import { describe, expect, it } from 'vitest'
import { errorFromResponse, parseErrorBody } from '@core/stt'
import { currentUtcDay, msUntilNextUtcDay, type UsageMeter } from '@shared/cloud'
import {
  describeLimit,
  formatAudioSeconds,
  formatResetTime,
  isPlanLimit,
  meterLabel,
  meterValue,
  parseLimitNotice,
  planStateLabel,
  planStateOf,
  trialDaysLeft,
  usageMeters,
  type LimitNotice
} from '@shared/limits'

const NOW = Date.UTC(2026, 8, 13, 12, 0, 0) // Sun 13 Sep 2026, noon UTC
const UPGRADE = 'https://murmur.app/account?upgrade=yearly'

/** A gateway refusal, as the contract spells it (internal/entitlements-contract.md). */
function refusal(overrides: Partial<Record<string, unknown>> = {}): string {
  return JSON.stringify({
    error: {
      type: 'murmur_gateway_error',
      code: 'quota_exceeded',
      message: "This week's 500 free words are used up.",
      limit: 'wordsPerWeek',
      plan: 'free',
      planState: 'free',
      used: 503,
      allowed: 500,
      resetsAt: NOW + 2 * 86_400_000,
      upgradeUrl: UPGRADE,
      ...overrides
    }
  })
}

describe('limit notices from the gateway', () => {
  it('are read off a refusal body and travel on the SttError', () => {
    const parsed = parseErrorBody(refusal())
    expect(parsed.code).toBe('quota_exceeded')
    expect(parsed.message).toBe("This week's 500 free words are used up.")
    expect(parsed.limit).toEqual<LimitNotice>({
      limit: 'wordsPerWeek',
      plan: 'free',
      planState: 'free',
      used: 503,
      allowed: 500,
      resetsAt: NOW + 2 * 86_400_000,
      upgradeUrl: UPGRADE,
      message: "This week's 500 free words are used up."
    })
    const err = errorFromResponse(429, refusal())
    expect(err.kind).toBe('rate-limit')
    expect(err.code).toBe('quota_exceeded')
    expect(err.limit?.limit).toBe('wordsPerWeek')
    expect(err.limit?.message).toBe(err.message)
    // A clip refusal is a 413 with no reset and no upgrade for a Pro account.
    const clip = errorFromResponse(
      413,
      refusal({
        code: 'clip_too_long',
        limit: 'maxClipSeconds',
        plan: 'pro',
        planState: 'pro',
        used: 640,
        allowed: 600,
        resetsAt: null,
        upgradeUrl: null
      })
    )
    expect(clip.limit).toMatchObject({ limit: 'maxClipSeconds', resetsAt: null, upgradeUrl: null })
  })

  it('leave ordinary errors alone', () => {
    expect(
      parseErrorBody('{"error":{"message":"boom","code":"upstream_error"}}').limit
    ).toBeUndefined()
    expect(parseErrorBody('not json').limit).toBeUndefined()
    expect(parseLimitNotice({ limit: 'somethingElse', used: 1 })).toBeNull()
    expect(parseLimitNotice(null)).toBeNull()
    expect(parseLimitNotice('wordsPerWeek')).toBeNull()
    // Missing details default sensibly; the tier falls back to the plan.
    expect(parseLimitNotice({ limit: 'dictationsPerDay', plan: 'pro' })).toEqual({
      limit: 'dictationsPerDay',
      plan: 'pro',
      planState: 'pro',
      used: 0,
      allowed: 0,
      resetsAt: null,
      upgradeUrl: null,
      message: ''
    })
  })

  it('tell a plan limit from a plain rate limit', () => {
    expect(isPlanLimit('wordsPerWeek')).toBe(true)
    expect(isPlanLimit('fairUseSttSecondsPerMonth')).toBe(true)
    expect(isPlanLimit('requestsPerMinute')).toBe(false)
  })
})

describe('limit wording', () => {
  const free = parseLimitNotice(JSON.parse(refusal()).error)!

  it('names the limit, the allowance and the reset', () => {
    const copy = describeLimit(free, NOW)
    expect(copy.title).toBe("This week's free words are used up")
    expect(copy.detail).toBe('500 words a week on the free plan · resets Tue 15 Sep')
    expect(copy.upgradeHelps).toBe(true)

    const day = describeLimit(
      { ...free, limit: 'dictationsPerDay', used: 12, allowed: 12, resetsAt: NOW + 5 * 3_600_000 },
      NOW
    )
    expect(day.title).toBe("Today's free dictations are used up")
    expect(day.detail.startsWith('12 dictations a day on the free plan · resets ')).toBe(true)
    // The same limit met at the formatting stage never loses text; the wording says so.
    expect(describeLimit({ ...free, limit: 'dictationsPerDay' }, NOW, 'formatting').title).toBe(
      "Inserted without formatting: today's free dictations are used up"
    )

    const minutes = describeLimit({ ...free, limit: 'sttSecondsPerWeek', allowed: 420 }, NOW)
    expect(minutes.title).toBe("This week's free minutes are used up")
    expect(minutes.detail).toBe('7 min of speech a week on the free plan · resets Tue 15 Sep')

    const clip = describeLimit(
      { ...free, limit: 'maxClipSeconds', used: 75, allowed: 60, resetsAt: null },
      NOW
    )
    expect(clip.title).toBe('That recording is too long for the free plan')
    expect(clip.detail).toBe('Clips up to 1 min on the free plan; up to 10 min on Pro')
    expect(clip.upgradeHelps).toBe(true)
  })

  it('does not sell Pro to Pro', () => {
    const pro: LimitNotice = {
      ...free,
      plan: 'pro',
      planState: 'pro',
      upgradeUrl: null,
      limit: 'sttSecondsPerMonth',
      used: 216_000,
      allowed: 216_000,
      resetsAt: Date.UTC(2026, 9, 1)
    }
    const cap = describeLimit(pro, NOW)
    expect(cap.title).toBe("This month's transcription has reached its fair-use cap")
    expect(cap.detail).toBe('60 h a month on Pro · resets on 1 Oct')
    expect(cap.upgradeHelps).toBe(false)

    const soft = describeLimit(
      { ...pro, limit: 'fairUseSttSecondsPerMonth', used: 108_500, allowed: 108_000 },
      NOW,
      'formatting'
    )
    expect(soft.title).toBe('Inserted without formatting: fair use reached for this month')
    expect(soft.detail).toBe(
      'Past 30 h of transcription a month the text is tidied by rules only · resets on 1 Oct'
    )
    expect(soft.upgradeHelps).toBe(false)

    const trialClip = describeLimit(
      {
        ...pro,
        planState: 'trial',
        limit: 'maxClipSeconds',
        used: 700,
        allowed: 600,
        resetsAt: null
      },
      NOW
    )
    expect(trialClip.title).toBe('That recording is too long')
    expect(trialClip.detail).toBe('Clips can be up to 10 min')
  })

  it('says when a limit comes back the way a person would', () => {
    expect(formatResetTime(NOW + 20_000, NOW)).toBe('in a moment')
    expect(formatResetTime(NOW + 41 * 60_000, NOW)).toBe('in 41 min')
    const later = NOW + 3 * 3_600_000
    const sameDay = new Date(later).getDate() === new Date(NOW).getDate()
    expect(formatResetTime(later, NOW)).toMatch(
      sameDay ? /^at \d{1,2}:\d{2} [ap]m$/ : /^tomorrow at /
    )
    expect(formatResetTime(NOW + 26 * 3_600_000, NOW)).toMatch(/^tomorrow at \d{1,2}:\d{2} [ap]m$/)
    expect(formatResetTime(NOW + 3 * 86_400_000, NOW)).toBe('Wed 16 Sep')
    expect(formatResetTime(NOW + 12 * 86_400_000, NOW)).toBe('on 25 Sep')
  })

  it('formats audio in the unit the allowance is read in', () => {
    expect(formatAudioSeconds(420)).toBe('7 min')
    expect(formatAudioSeconds(7_200)).toBe('120 min')
    expect(formatAudioSeconds(30)).toBe('<1 min')
    expect(formatAudioSeconds(0)).toBe('0 min')
    expect(formatAudioSeconds(108_000)).toBe('30 h')
    expect(formatAudioSeconds(9_000, 108_000)).toBe('2.5 h')
    expect(formatAudioSeconds(45_000, 216_000)).toBe('13 h')
  })

  it('describes the Account page meters', () => {
    const meters: UsageMeter[] = [
      { limit: 'wordsPerWeek', used: 312, allowed: 500, exceeded: false, resetsAt: NOW },
      { limit: 'sttSecondsPerWeek', used: 95, allowed: 420, exceeded: false, resetsAt: NOW },
      { limit: 'dictationsPerDay', used: 3, allowed: 12, exceeded: false, resetsAt: NOW },
      { limit: 'maxClipSeconds', used: 0, allowed: 60, exceeded: false, resetsAt: NOW },
      { limit: 'requestsPerMinute', used: 1, allowed: 20, exceeded: false, resetsAt: NOW },
      { limit: 'llmTokensPerMonth', used: 12_400, allowed: 500_000, exceeded: false, resetsAt: NOW }
    ]
    expect(meterValue(meters[0])).toBe('312 of 500 words')
    expect(meterValue(meters[1])).toBe('2 of 7 min')
    expect(meterValue(meters[2])).toBe('3 of 12')
    expect(meterValue(meters[3])).toBe('up to 1 min a clip')
    expect(meterValue(meters[5])).toBe('12k of 500k tokens')
    expect(meterValue({ limit: 'llmTokensPerMonth', used: 2_100_000, allowed: 25_000_000 })).toBe(
      '2.1M of 25M tokens'
    )
    expect(meterValue({ limit: 'llmTokensPerMonth', used: 980, allowed: 500_000 })).toBe(
      '980 of 500k tokens'
    )
    expect(meterLabel('wordsPerWeek')).toBe('Words this week')
    expect(meterLabel('fairUseSttSecondsPerMonth')).toBe('Fair use this month')
    // Per-request limits are not usage; they get no row.
    expect(usageMeters(meters).map((m) => m.limit)).toEqual([
      'wordsPerWeek',
      'sttSecondsPerWeek',
      'dictationsPerDay',
      'llmTokensPerMonth'
    ])
    expect(usageMeters(undefined)).toEqual([])
  })
})

describe('plan state', () => {
  it('comes from the instance when it reports one, otherwise from the tier', () => {
    expect(planStateOf({ plan: 'pro', planState: 'trial' })).toBe('trial')
    expect(planStateOf({ plan: 'pro' })).toBe('pro')
    expect(planStateOf({ plan: 'free' })).toBe('free')
    expect(planStateOf(null, { plan: 'pro', planState: 'trial' })).toBe('trial')
    expect(planStateOf(null, { plan: 'free' })).toBe('free')
    expect(planStateOf(null, null)).toBe('free')
    expect(planStateLabel('trial')).toBe('Pro trial')
    expect(planStateLabel('pro')).toBe('Pro')
    expect(planStateLabel('free')).toBe('Free')
  })

  it('counts trial days the way the contract does', () => {
    expect(trialDaysLeft(NOW + 14 * 86_400_000, NOW)).toBe(14)
    expect(trialDaysLeft(NOW + 1, NOW)).toBe(1)
    expect(trialDaysLeft(NOW - 5, NOW)).toBe(0)
    expect(trialDaysLeft(null, NOW)).toBe(0)
    expect(trialDaysLeft(undefined, NOW)).toBe(0)
  })

  it('knows the UTC day the status is asked for and when it turns', () => {
    expect(currentUtcDay(NOW)).toBe('2026-09-13')
    expect(msUntilNextUtcDay(NOW)).toBe(12 * 3_600_000)
    expect(currentUtcDay(NOW + msUntilNextUtcDay(NOW))).toBe('2026-09-14')
    expect(msUntilNextUtcDay(Date.UTC(2026, 8, 13, 23, 59, 59, 999))).toBe(1)
  })
})

File: apps/android/app/src/test/java/app/murmur/android/LimitsTest.kt
package app.murmur.android

import app.murmur.android.cloud.InferenceStatusDto
import app.murmur.android.cloud.UsageMeterDto
import app.murmur.android.cloud.UserDto
import app.murmur.android.inference.InferenceRouting
import app.murmur.android.inference.LimitNotice
import app.murmur.android.inference.LimitStage
import app.murmur.android.inference.Limits
import app.murmur.android.inference.PlanActions
import app.murmur.android.settings.InferenceSource
import app.murmur.android.ui.InferenceView
import app.murmur.android.stt.SttErrorKind
import app.murmur.android.stt.errorFromResponse
import app.murmur.android.stt.parseErrorBody
import app.murmur.android.text.FormatOutcome
import app.murmur.android.text.FormatResult
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.Calendar
import java.util.TimeZone

private const val UPGRADE = "https://murmur.app/account?upgrade=yearly"
private const val ACCOUNT = "https://murmur.app/account"

/** Sun 13 Sep 2026, noon UTC. */
private val NOW: Long = Calendar.getInstance(TimeZone.getTimeZone("UTC")).apply {
    clear()
    set(2026, Calendar.SEPTEMBER, 13, 12, 0, 0)
}.timeInMillis
private const val DAY = 86_400_000L

/** A gateway refusal, as the contract spells it (internal/entitlements-contract.md). */
private fun refusal(vararg overrides: Pair<String, Any?>): String {
    val error = JSONObject()
        .put("type", "murmur_gateway_error")
        .put("code", "quota_exceeded")
        .put("message", "This week's 500 free words are used up.")
        .put("limit", "wordsPerWeek")
        .put("plan", "free")
        .put("planState", "free")
        .put("used", 503)
        .put("allowed", 500)
        .put("resetsAt", NOW + 2 * DAY)
        .put("upgradeUrl", UPGRADE)
        .put("accountUrl", ACCOUNT)
    for ((k, v) in overrides) if (v == null) error.put(k, JSONObject.NULL) else error.put(k, v)
    return JSONObject().put("error", error).toString()
}

/** The limit notice and the wording built on it (port of shared/limits.ts and its tests). */
class LimitsTest {

    @Test
    fun `a refusal body yields a notice that travels on the exception`() {
        val parsed = parseErrorBody(refusal())
        assertEquals("quota_exceeded", parsed.code)
        assertEquals(
            LimitNotice("wordsPerWeek", "free", "free", 503.0, 500.0, NOW + 2 * DAY, UPGRADE, ACCOUNT, "This week's 500 free words are used up."),
            parsed.limit
        )
        val e = errorFromResponse(429, refusal())
        assertEquals(SttErrorKind.RATE_LIMIT, e.kind)
        assertEquals("wordsPerWeek", e.limit?.limit)
        assertEquals(e.message, e.limit?.message)
        assertNotNull(e.planLimit)
        assertEquals(e.message, e.friendly())

        // A clip refusal is a 413 with no reset and no upgrade for a Pro account.
        val clip = errorFromResponse(
            413,
            refusal("code" to "clip_too_long", "limit" to "maxClipSeconds", "plan" to "pro", "planState" to "pro", "used" to 640, "allowed" to 600, "resetsAt" to null, "upgradeUrl" to null)
        )
        assertEquals("maxClipSeconds", clip.limit?.limit)
        assertNull(clip.limit?.resetsAt)
        assertNull(clip.limit?.upgradeUrl)
        assertEquals(ACCOUNT, clip.limit?.accountUrl)
        // An instance without a site URL sends null for both pages.
        val siteless = errorFromResponse(429, refusal("upgradeUrl" to null, "accountUrl" to null))
        assertNull(siteless.limit?.upgradeUrl)
        assertNull(siteless.limit?.accountUrl)

        // A rate limit carries a notice but is not a plan limit: the ordinary error treatment.
        val rate = errorFromResponse(429, refusal("code" to "rate_limited", "limit" to "requestsPerMinute", "used" to 20, "allowed" to 20))
        assertEquals("requestsPerMinute", rate.limit?.limit)
        assertNull(rate.planLimit)
    }

    @Test
    fun `ordinary errors carry no notice`() {
        assertNull(parseErrorBody("""{"error":{"message":"boom","code":"upstream_error"}}""").limit)
        assertNull(parseErrorBody("not json").limit)
        assertNull(LimitNotice.fromJson(JSONObject().put("limit", "somethingElse")))
        assertNull(LimitNotice.fromJson(null))
        // Missing details default sensibly; the tier falls back to the plan.
        assertEquals(
            LimitNotice("dictationsPerDay", "pro", "pro", 0.0, 0.0, null, null, null, ""),
            LimitNotice.fromJson(JSONObject().put("limit", "dictationsPerDay").put("plan", "pro"))
        )
    }

    @Test
    fun `a format answer skipped for fair use carries the limit that paused the model`() {
        val body = JSONObject(
            """{"text":"Hello there","pressEnter":false,"status":{"outcome":"skipped","detail":"fair use"},"llmMs":0,"stages":[],
               "limit":{"limit":"fairUseSttSecondsPerMonth","plan":"pro","planState":"pro","used":108500,"allowed":108000,"resetsAt":${NOW + 10 * DAY},"upgradeUrl":null}}"""
        )
        val result = FormatResult.fromJson(body)
        assertEquals(FormatOutcome.SKIPPED, result.status.outcome)
        assertEquals("fairUseSttSecondsPerMonth", result.limit?.limit)
        assertEquals("fair use", result.limit?.message)
        assertNull(FormatResult.fromJson(JSONObject("""{"text":"x","status":{"outcome":"used"}}""")).limit)
    }

    @Test
    fun `the wording names the limit, the allowance and the reset`() {
        val free = LimitNotice.fromJson(JSONObject(refusal()).getJSONObject("error"))!!
        val copy = Limits.describe(free, NOW)
        assertEquals("This week's free words are used up", copy.title)
        assertEquals("500 words a week on the free plan · resets Tue 15 Sep", copy.detail)
        assertTrue(copy.upgradeHelps)

        val day = Limits.describe(free.copy(limit = "dictationsPerDay", used = 12.0, allowed = 12.0, resetsAt = NOW + 5 * 3_600_000), NOW)
        assertEquals("Today's free dictations are used up", day.title)
        assertTrue(day.detail, day.detail.startsWith("12 dictations a day on the free plan · resets "))
        val unformatted = Limits.describe(free.copy(limit = "dictationsPerDay"), NOW, LimitStage.FORMATTING)
        assertEquals("Inserted without formatting", unformatted.title)
        assertEquals("Today's free dictations are used up · resets Tue 15 Sep", unformatted.detail)

        val minutes = Limits.describe(free.copy(limit = "sttSecondsPerWeek", allowed = 420.0), NOW)
        assertEquals("This week's free minutes are used up", minutes.title)
        assertEquals("7 min of speech a week on the free plan · resets Tue 15 Sep", minutes.detail)

        val clip = Limits.describe(free.copy(limit = "maxClipSeconds", used = 75.0, allowed = 60.0, resetsAt = null), NOW)
        assertEquals("That recording is too long for the free plan", clip.title)
        assertEquals("Clips up to 1 min on the free plan; up to 10 min on Pro", clip.detail)
    }

    @Test
    fun `Pro is not sold Pro`() {
        val pro = LimitNotice("sttSecondsPerMonth", "pro", "pro", 216_000.0, 216_000.0, utc(2026, Calendar.OCTOBER, 1), null, ACCOUNT, "")
        val cap = Limits.describe(pro, NOW)
        assertEquals("This month's fair-use cap is reached", cap.title)
        assertEquals("60 h a month on Pro · resets on 1 Oct", cap.detail)
        assertFalse(cap.upgradeHelps)

        val soft = Limits.describe(pro.copy(limit = "fairUseSttSecondsPerMonth", used = 108_500.0, allowed = 108_000.0), NOW, LimitStage.FORMATTING)
        assertEquals("Inserted without formatting", soft.title)
        assertEquals("Fair use reached for this month · resets on 1 Oct", soft.detail)
        assertFalse(soft.upgradeHelps)
        val paused = Limits.describe(pro.copy(limit = "fairUseSttSecondsPerMonth", used = 108_500.0, allowed = 108_000.0), NOW)
        assertEquals("Fair use reached for this month", paused.title)
        assertEquals("Past 30 h of transcription a month the text is tidied by rules only · resets on 1 Oct", paused.detail)

        val trialClip = Limits.describe(pro.copy(planState = "trial", limit = "maxClipSeconds", used = 700.0, allowed = 600.0, resetsAt = null), NOW)
        assertEquals("That recording is too long", trialClip.title)
        assertEquals("Clips can be up to 10 min", trialClip.detail)
    }

    @Test
    fun `resets are said the way a person would`() {
        assertEquals("in a moment", Limits.formatResetTime(NOW + 20_000, NOW))
        assertEquals("in 41 min", Limits.formatResetTime(NOW + 41 * 60_000, NOW))
        assertTrue(Limits.formatResetTime(NOW + 26 * 3_600_000, NOW).matches(Regex("tomorrow at \\d{1,2}:\\d{2} [ap]m")))
        assertEquals("Wed 16 Sep", Limits.formatResetTime(NOW + 3 * DAY, NOW))
        assertEquals("on 25 Sep", Limits.formatResetTime(NOW + 12 * DAY, NOW))
    }

    @Test
    fun `figures and meters read in their natural unit`() {
        assertEquals("7 min", Limits.formatAudioSeconds(420.0))
        assertEquals("120 min", Limits.formatAudioSeconds(7_200.0))
        assertEquals("<1 min", Limits.formatAudioSeconds(30.0))
        assertEquals("30 h", Limits.formatAudioSeconds(108_000.0))
        assertEquals("2.5 h", Limits.formatAudioSeconds(9_000.0, 108_000.0))
        assertEquals("312 of 500 words", Limits.meterValue("wordsPerWeek", 312.0, 500.0))
        assertEquals("2 of 7 min", Limits.meterValue("sttSecondsPerWeek", 95.0, 420.0))
        assertEquals("3 of 12", Limits.meterValue("dictationsPerDay", 3.0, 12.0))
        assertEquals("12k of 500k tokens", Limits.meterValue("llmTokensPerMonth", 12_400.0, 500_000.0))
        assertEquals("2.1M of 25M tokens", Limits.meterValue("llmTokensPerMonth", 2_100_000.0, 25_000_000.0))
        assertEquals("980 of 500k tokens", Limits.meterValue("llmTokensPerMonth", 980.0, 500_000.0))
        assertEquals("Words this week", Limits.meterLabel("wordsPerWeek"))
        val meters = listOf(
            UsageMeterDto("wordsPerWeek", 312.0, 500.0),
            UsageMeterDto("maxClipSeconds", 0.0, 60.0),
            UsageMeterDto("requestsPerMinute", 1.0, 20.0),
            UsageMeterDto("llmTokensPerMonth", 1.0, 500_000.0),
            UsageMeterDto("somethingNew", 1.0, 2.0)
        )
        assertEquals(listOf("wordsPerWeek", "llmTokensPerMonth"), Limits.usageMeters(meters).map { it.limit })
    }

    @Test
    fun `plan state comes from the instance, otherwise from the tier`() {
        assertEquals("trial", Limits.planStateOf(InferenceStatusDto(plan = "pro", planState = "trial"), null))
        assertEquals("pro", Limits.planStateOf(InferenceStatusDto(plan = "pro"), null))
        assertEquals("free", Limits.planStateOf(InferenceStatusDto(plan = "free"), null))
        assertEquals("trial", Limits.planStateOf(null, UserDto("u", "c", plan = "pro", planState = "trial")))
        assertEquals("free", Limits.planStateOf(null, null))
        assertEquals("Pro trial", Limits.planStateLabel("trial"))
        assertEquals("Pro trial", Limits.planTitle("trial"))
        assertEquals("Free plan", Limits.planTitle("free"))
        assertEquals(14, Limits.trialDaysLeft((NOW + 14 * DAY).toDouble(), NOW))
        assertEquals(1, Limits.trialDaysLeft((NOW + 1).toDouble(), NOW))
        assertEquals(0, Limits.trialDaysLeft((NOW - 5).toDouble(), NOW))
        assertEquals(0, Limits.trialDaysLeft(null, NOW))
    }

    @Test
    fun `Upgrade and Manage plan go only to whom they apply, and only with a page to open`() {
        assertEquals(PlanActions(UPGRADE, ACCOUNT), Limits.planActions("trial", UPGRADE, ACCOUNT))
        assertEquals(PlanActions(UPGRADE, null), Limits.planActions("free", UPGRADE, ACCOUNT))
        assertEquals(PlanActions(null, ACCOUNT), Limits.planActions("pro", null, ACCOUNT))
        // No site URL on the instance: nothing to open, so no buttons at all.
        for (state in listOf("trial", "free", "pro")) assertEquals(PlanActions(null, null), Limits.planActions(state, null, null))
        assertEquals(PlanActions(null, null), Limits.planActions("pro", "", ""))
        // The view folds the status in; an older instance without the field is the same as null.
        val view = InferenceView(
            cloudEnabled = true, managedAvailable = true, routing = InferenceRouting(InferenceSource.MURMUR, InferenceSource.MURMUR),
            signedIn = true, status = InferenceStatusDto(plan = "pro", planState = "pro"), plan = "pro", planState = "pro",
            trialDaysLeft = 0, sttReady = true, llmReady = true
        )
        assertEquals(PlanActions(null, null), view.planActions)
        assertEquals(PlanActions(null, ACCOUNT), view.copy(status = view.status?.copy(accountUrl = ACCOUNT)).planActions)
    }

    @Test
    fun `the status is asked for the UTC day, and again when it turns`() {
        assertEquals("2026-09-13", InferenceStatusDto.currentUtcDay(NOW))
        assertEquals(12 * 3_600_000L, InferenceStatusDto.msUntilNextUtcDay(NOW))
        assertEquals("2026-09-14", InferenceStatusDto.currentUtcDay(NOW + InferenceStatusDto.msUntilNextUtcDay(NOW)))
        assertEquals(1L, InferenceStatusDto.msUntilNextUtcDay(NOW + 12 * 3_600_000L - 1))
    }

    private fun utc(year: Int, month: Int, day: Int): Long =
        Calendar.getInstance(TimeZone.getTimeZone("UTC")).apply { clear(); set(year, month, day) }.timeInMillis
}
